src/cpp/shared/combat/army.cpp

```diff
@@ -614,6 +650,293 @@ void army::Walk(signed int dir, int last, int notFirst) {
   }
 }
 
+int army::FindPath(int knownHex, int targHex, int speed, int flying, int flag) {
+  int res;
+  std::vector<int> obstacleHexes;
+
+  if(!IsAICombatTurn() && CreatureHasAttribute(this->creatureIdx, JUMPER) && gIronfistExtra.combat.stack.abilityCounter[this]) {
+    for (int i = 0; i < 117; i++) {
+      if (gpCombatManager->combatGrid[i].isBlocked != 0 && !IsCastleWall(i)) {
+        obstacleHexes.push_back(i);
+        gpCombatManager->combatGrid[i].isBlocked = 0;
+      }
+    }
+  }
+  res = this->FindPath_orig(this->occupiedHex, targHex, this->creature.speed, 0, flag);
+
+  // pretending we don't see obstacles at all
+  // this does nothing for creatures that don't ignore obstacles
+  for (auto i : obstacleHexes)
+    gpCombatManager->combatGrid[i].isBlocked = 1;
+
+  return res;
+}
+
+int army::ValidPath(int hex, int flag) {
+  if (ValidHex(hex)) {
+    if (this->creature.creature_flags & FLYER) {
+      return this->ValidFlight(hex, flag);
+    } else if (this->FindPath(this->occupiedHex, hex, this->creature.speed, 0, flag)) {
+      this->targetHex = hex;
+      return 1;
+    } else {
+      return 0;
+    }
+  }
+  return 0;
+}
+
+#pragma pack(push, 1)
+struct PathfindingInfo {
+  char field_0;
+  char field_1;
+  __int16 field_2;
+  char field_4;
+  int field_5;
+};
+#pragma pack(pop)
+
+#pragma pack(push,1)
+class searchArray {
+public:
+	int field_0;
+	int field_4;
+	int field_8;
+	char _1[8];
+	PathfindingInfo mainDataStructure[1024];
+	PathfindingInfo *field_2414;
+	int field_2418;
+	int field_241C[63];
+	searchArray();
+};
+#pragma pack(pop)
+
+void army::ArcJump(int fromHex, int toHex) {
```

Comment:
Where did this code come from? Portions of it look decompiled, and I can't read it.

Is it possible to make this share more code with ArcShot?

Reply:
It's from catapult code. They influence different things and I can't think of anything that makes it reuse the same code without making it harder to read